src/git_indexer.h

```diff
@@ -10,34 +10,79 @@
 
 #include <string>
 #include "src/proto/config.pb.h"
+#include "src/smart_git.h"
+#include "src/lib/threadsafe_progress_indicator.h"
+#include "src/lib/thread_queue.h"
 
 class code_searcher;
 class git_repository;
 class git_tree;
 struct indexed_tree;
 
+// We walk all repos and create a vector of pre_indexed_file(s)
+// We then sort that vector by score (and repo name)
+// In that way, low scoring files are indexed at the back of a chunk,
+// or are contained mostly in the last chunks, so higher ranked files
+// appear sooner in search results, since we search by chunk.
+struct pre_indexed_file {
+    const indexed_tree *tree;
+    std::string  repopath;
+    std::string  path;
+    int score;
+    git_repository *repo;
+    git_oid *oid;
+};
+
+// used to walk a repos folders/trees in parallel using threads
+struct tree_to_walk {
+    int id;
+    std::string prefix;
+    std::string repopath;
+    bool walk_submodules;
+    string submodule_prefix;
+    const indexed_tree *idx_tree;
+    git_tree *tree;
+    git_repository *repo;
```

Comment:
We eat the space overhead of storing a pointer to a git repo with every tree and file, since even with millions of files and trees it's only 10s of MB, and it's slower (in experiments) than having git_repo in a map and then fetching them out when needed.